plugin/storage/es/factory.go

```diff
@@ -236,8 +266,26 @@ var _ io.Closer = (*Factory)(nil)
 
 // Close closes the resources held by the factory
 func (f *Factory) Close() error {
+	var errs []error
+	for _, w := range f.watchers {
+		errs = append(errs, w.Close())
+	}
 	if cfg := f.Options.Get(archiveNamespace); cfg != nil {
 		cfg.TLS.Close()
 	}
-	return f.Options.GetPrimary().TLS.Close()
+	errs = append(errs, f.Options.GetPrimary().TLS.Close())
+	return errors.Join(errs...)
+}
+
+func (f *Factory) onPasswordChange(config config.ClientBuilder, esClient es.Client) {
+	b, err := os.ReadFile(filepath.Clean(config.GetPasswordFilePath()))
+	if err == nil {
+		newPassword := strings.TrimRight(string(b), "\r\n")
+		config.SetPassword(newPassword)
+		esClient.SetBasicAuth(config.GetUsername(), newPassword)
```

Comment:
is `esClient.SetBasicAuth` method reentrant? That is, would the take effect after the client is already created? It would be good to add a unit test that validates it (assuming we can mock the ES server with a test server, which iirc we do somewhere by using a fake http.RoundTripper)

Reply:
From my study of the codebase of `github.com/olivere/elastic`, yes.

You can see the implementation of `esClient.SetBasicAuth` [here](https://github.com/jaegertracing/jaeger/pull/4342/files#diff-178ba3fec75295e1f7b9539382116134f29d80c7e52305cf8a87b4ffc8fdeb80R84-R87).

Basically what it does is what `github.com/olivere/elastic` does to set username and password on ES client: [get a ClientOptionFunc from calling SetBasicAuth](https://github.com/olivere/elastic/blob/29ee98974cf1984dfecf53ef772d721fb97cb0b9/client.go#L466) then [apply that ClientOptionFunc on ES client](https://github.com/olivere/elastic/blob/release-branch.v7/client.go#L347).

I'll add a test for this.